File: tests/ten_runtime/integration/python/sync_stop_before_deinit_python/sync_stop_before_deinit_python_app/ten_packages/extension/test_extension_1/extension.py
#
# Copyright © 2025 Agora
# This file is part of TEN Framework, an open source project.
# Licensed under the Apache License, Version 2.0, with certain conditions.
# Refer to the "LICENSE" file in the root directory for more information.
#
# test_extension_1: orchestrator living in the predefined "default" graph.
#
# Flow:
#   1. on_start: launch a dynamic graph with sync_stop_before_deinit=True
#      (contains test_extension_2 and test_extension_3), then immediately
#      send stop_graph to trigger the on_stop phase.
#   2. During the on_start await for stop_graph, the asyncio event loop can
#      concurrently dispatch on_cmd("on_stop_notify") coming from
#      test_extension_2 (which sleeps 2s then sends the notification).
#   3. We reply to the result of on_stop_notify so test_extension_2 can call
#      on_stop_done.
#   4. Once stop_graph completes AND the notification is received, we reply to
#      the client.  The client assertion proves the message was delivered while
#      the dynamic graph was still in its stop phase.
#

import json
from ten_runtime import (
    AsyncExtension,
    AsyncTenEnv,
    Cmd,
    StatusCode,
    CmdResult,
    Loc,
    StartGraphCmd,
    StopGraphCmd,
)


class TestExtension1(AsyncExtension):
    def __init__(self, name: str) -> None:
        super().__init__(name)
        self._test_cmd: Cmd | None = None
        self._stop_graph_done: bool = False
        self._got_on_stop_notify: bool = False

    async def on_start(self, ten_env: AsyncTenEnv) -> None:
        # Build the dynamic graph with sync_stop_before_deinit enabled.
        start_graph_cmd = StartGraphCmd.create()
        start_graph_cmd.set_dests([Loc("")])
        start_graph_cmd.set_sync_stop_before_deinit(True)

        # extension_2 (slow stopper) exposes on_stop_notify as cmd_out so the
        # message can cross the graph boundary and reach test_extension_1.
        graph_json = {
            "nodes": [
                {
                    "type": "extension",
                    "name": "test_extension_2",
                    "addon": "test_extension_2",
                    "extension_group": "group_2",
                },
                {
                    "type": "extension",
                    "name": "test_extension_3",
                    "addon": "test_extension_3",
                    "extension_group": "group_3",
                },
            ],
            "exposed_messages": [
                {
                    "type": "cmd_out",
                    "name": "on_stop_notify",
                    "extension": "test_extension_2",
                }
            ],
        }
        start_graph_cmd.set_graph_from_json(json.dumps(graph_json))

        cmd_result, error = await ten_env.send_cmd(start_graph_cmd)
        if error is not None:
            ten_env.log_error(f"start_graph failed: {error}")
            return
        if cmd_result is None:
            ten_env.log_error("start_graph cmd_result is None")
            return

        graph_id, _ = cmd_result.get_property_string("graph_id")
        ten_env.log_info(f"Dynamic graph started, id={graph_id}")

        # Immediately stop the dynamic graph to trigger on_stop for both
        # extensions.  While we await this, the asyncio event loop can
        # concurrently run on_cmd("on_stop_notify").
        stop_graph_cmd = StopGraphCmd.create()
        stop_graph_cmd.set_dests([Loc("")])
        stop_graph_cmd.set_graph_id(graph_id)

        _, error = await ten_env.send_cmd(stop_graph_cmd)
        if error is not None:
            ten_env.log_error(f"stop_graph failed: {error}")
            return

        ten_env.log_info("stop_graph done")
        self._stop_graph_done = True
        await self._try_reply(ten_env)

    async def on_cmd(self, ten_env: AsyncTenEnv, cmd: Cmd) -> None:
        cmd_name = cmd.get_name()

        if cmd_name == "test":
            self._test_cmd = cmd
            await self._try_reply(ten_env)

        elif cmd_name == "on_stop_notify":
            ten_env.log_info("Received on_stop_notify from dynamic graph")
            self._got_on_stop_notify = True

            # Must reply so test_extension_2's send_cmd callback fires and it
            # can call on_stop_done.
            await ten_env.return_result(CmdResult.create(StatusCode.OK, cmd))

            await self._try_reply(ten_env)

        else:
            ten_env.log_error(f"Unexpected cmd: {cmd_name}")

    async def _try_reply(self, ten_env: AsyncTenEnv) -> None:
        if (
            self._test_cmd is None
            or not self._stop_graph_done
            or not self._got_on_stop_notify
        ):
            return

        ten_env.log_info("All conditions met, replying to client")
        cmd_result = CmdResult.create(StatusCode.OK, self._test_cmd)
        cmd_result.set_property_string(
            "detail", json.dumps({"id": 1, "name": "a"})
        )
        await ten_env.return_result(cmd_result)
        self._test_cmd = None

File: tests/ten_runtime/integration/go/sync_stop_before_deinit_go/sync_stop_before_deinit_go_app/ten_packages/extension/test_extension_1/extension.go
//
// Copyright © 2025 Agora
// This file is part of TEN Framework, an open source project.
// Licensed under the Apache License, Version 2.0, with certain conditions.
// Refer to the "LICENSE" file in the root directory for more information.
//
// test_extension_1: orchestrator living in the predefined "default" graph.
//
// Flow:
//   1. On start: launch a dynamic graph with sync_stop_before_deinit=true
//      (contains test_extension_2 and test_extension_3).
//   2. Immediately stop the dynamic graph so both extensions enter on_stop.
//   3. test_extension_2 sleeps 2 s in on_stop, then sends "on_stop_notify".
//   4. We receive "on_stop_notify" and return a result so test_extension_2
//      can proceed to call on_stop_done.
//   5. Once stop_graph is done AND the notification is received, reply to the
//      client with OK.  The client's assertion proves the message was
//      delivered while the graph was still in its stop phase.
//

package test_extension_1

import (
	"encoding/json"

	ten "ten_framework/ten_runtime"
)

type testExtension1 struct {
	ten.DefaultExtension

	newGraphID          string
	testCmd             ten.Cmd
	stopGraphDone       bool
	gotOnStopNotify     bool
}

func (ext *testExtension1) OnStart(tenEnv ten.TenEnv) {
	startGraphCmd, _ := ten.NewStartGraphCmd()
	startGraphCmd.SetDests(ten.Loc{
		AppURI:        ten.Ptr(""),
		GraphID:       nil,
		ExtensionName: nil,
	})

	// Enable sync_stop_before_deinit so every extension in the dynamic graph
	// must call on_stop_done before any of them is allowed to proceed to
	// on_deinit.
	if err := startGraphCmd.SetSyncStopBeforeDeinit(true); err != nil {
		panic("Failed to set sync_stop_before_deinit: " + err.Error())
	}

	// Dynamic graph: extension_2 (slow stopper) and extension_3 (fast stopper).
	// extension_2 exposes on_stop_notify as cmd_out so the message can cross
	// the graph boundary and reach test_extension_1.
	graphJSON := `{
		"nodes": [{
			"type": "extension",
			"name": "test_extension_2",
			"addon": "test_extension_2",
			"extension_group": "group_2"
		}, {
			"type": "extension",
			"name": "test_extension_3",
			"addon": "test_extension_3",
			"extension_group": "group_3"
		}],
		"exposed_messages": [{
			"type": "cmd_out",
			"name": "on_stop_notify",
			"extension": "test_extension_2"
		}]
	}`

	if err := startGraphCmd.SetGraphFromJSONBytes([]byte(graphJSON)); err != nil {
		panic("Failed to set graph JSON: " + err.Error())
	}

	tenEnv.SendCmd(
		startGraphCmd,
		func(tenEnv ten.TenEnv, cmdResult ten.CmdResult, err error) {
			if err != nil {
				panic("Failed to start graph: " + err.Error())
			}

			statusCode, _ := cmdResult.GetStatusCode()
			if statusCode != ten.StatusCodeOk {
				panic("start_graph command failed")
			}

			newGraphID, _ := cmdResult.GetPropertyString("graph_id")
			ext.newGraphID = newGraphID
			tenEnv.LogInfo("Dynamic graph started, id=" + newGraphID)

			// Immediately stop the dynamic graph to trigger on_stop for both
			// extensions.
			stopGraphCmd, _ := ten.NewStopGraphCmd()
			stopGraphCmd.SetDests(ten.Loc{
				AppURI:        ten.Ptr(""),
				GraphID:       nil,
				ExtensionName: nil,
			})
			stopGraphCmd.SetGraphID(newGraphID)

			tenEnv.SendCmd(
				stopGraphCmd,
				func(tenEnv ten.TenEnv, _ ten.CmdResult, err error) {
					if err != nil {
						tenEnv.LogError("stop_graph failed: " + err.Error())
					}
					tenEnv.LogInfo("stop_graph done")
					ext.stopGraphDone = true
					ext.tryReply(tenEnv)
				},
			)
		},
	)

	tenEnv.OnStartDone()
}

func (ext *testExtension1) OnCmd(tenEnv ten.TenEnv, cmd ten.Cmd) {
	cmdName, _ := cmd.GetName()

	switch cmdName {
	case "test":
		ext.testCmd = cmd
		ext.tryReply(tenEnv)

	case "on_stop_notify":
		tenEnv.LogInfo("Received on_stop_notify from dynamic graph")
		ext.gotOnStopNotify = true

		// Reply so test_extension_2's send_cmd callback fires and it can call
		// on_stop_done.
		result, _ := ten.NewCmdResult(ten.StatusCodeOk, cmd)
		tenEnv.ReturnResult(result, nil)

		ext.tryReply(tenEnv)

	default:
		tenEnv.LogError("Unexpected cmd: " + cmdName)
	}
}

func (ext *testExtension1) tryReply(tenEnv ten.TenEnv) {
	if ext.testCmd == nil || !ext.stopGraphDone || !ext.gotOnStopNotify {
		return
	}

	tenEnv.LogInfo("All conditions met, replying to client")

	cmdResult, _ := ten.NewCmdResult(ten.StatusCodeOk, ext.testCmd)
	detail := map[string]interface{}{
		"id":   1,
		"name": "a",
	}
	detailBytes, _ := json.Marshal(detail)
	cmdResult.SetPropertyString("detail", string(detailBytes))

	tenEnv.ReturnResult(cmdResult, nil)
	ext.testCmd = nil
}

func newTestExtension1(name string) ten.Extension {
	return &testExtension1{}
}

func init() {
	if err := ten.RegisterAddonAsExtension(
		"test_extension_1",
		ten.NewDefaultExtensionAddon(newTestExtension1),
	); err != nil {
		panic("Failed to register addon: " + err.Error())
	}
}
